Autofill change to sorting algorithm for displaying logins

amddg44 opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.

⚠️ If you're an external contributor, please file an issue first before working on a PR, as we can't guarantee that we will accept your changes if they haven't been discussed ahead of time. Thanks!
-->

Task/Issue URL: https://app.asana.com/0/0/1203927830097167/f
Tech Design URL:
CC: @THISISDINOSAUR 

**Description**:
Changes to Autofill logins management screen taking subdomains and ports into account

<!--
Tagging instructions
If this PR isn't ready to be merged for whatever reason it should be marked with the `DO NOT MERGE` label (particularly if it's a draft)
If it's pending Product Review/PFR, please add the `Pending Product Review` label.

If at any point it isn't actively being worked on/ready for review/otherwise moving forward (besides the above PR/PFR exception) strongly consider closing it (or not opening it in the first place). If you decide not to close it, make sure it's labelled to make it clear the PRs state and comment with more information.
-->

**Steps to test this PR**:

**URL store changes** https://app.asana.com/0/0/1203927830097159/f, https://app.asana.com/0/1201645688934642/1203927830097162/f
1.  Go to https://fill.dev/form/login-simple
2. Login and store the credentials
3. Check the website URL saved is `fill.dev`
4. Add a new login manually and set the Website URL to something like `https://www.example.com?params=123`
5. Save and check the website URL saved is `www.example.com`
6. Add a new login manually with a port, setting the Website URL to something like `https://www.example.com:123`
7. Save and check the website URL saved is `www.example.com:123`

**URL matching changes** 
1. Add a login for `fill.dev`
2. Visit `fill.dev/form/login-simple` and confirms it prompts to autofill as normal
3. Open Autofill login list screen; verify it is listed as a suggested site
4. Edit the login and append a port to the URL (e.g., `fill.dev:9000`)
6. Visit the login page again (refreshing if required); verify you are not prompted to autofill (as ports don't match)
7. Open Autofill login list screen; confrim it is not listed as a suggested site

**URL sorting changes** https://app.asana.com/0/0/1203927830097167/f
1. Manually add the login entries detailed in https://app.asana.com/0/0/1203927830097167/f
2. In the Autofill login screen, confirm the logins are sorted in the same order as in the task
3. Add any other combinations of titles, domains and subdomains to test the sort order matches expectations 

**Leave title blank by default changes** https://app.asana.com/0/0/1203927830097163/f
1. Create a new login, setting only the Website URL, leaving the title blank. 
2. Once saved the title should be the Website URL
3. Tap to Edit the saved login and the title should display placeholder text 'Title'
4. Leave the title blank and save. Confirm the title still displays the Website URL and in edit mode still displays placeholder text 'Title'
5. Edit the login and set a title. Save and confirm the login is displaying the newly set title.
6. Save a new login via a website login save prompt. Tap to view the saved login
7. Confirm the title is displaying placeholder text 'Title' by default when in edit mode8. 

**Change to Favicon behavior** https://app.asana.com/0/0/1204031392481537/f
1. Create a login with a title that is different to the domain of the website
2. Confirm the placeholder letter is the first letter of the title, not of the domain
3. Create a login with a subdomain, but no title
4. Confirm the placeholder letter is the first letter of the domain, not of the subdomain
5. Confirm that otherwise favicons behave as before

**Changes to login prompt screen to illustrate partial subdomain matches** https://app.asana.com/0/0/1203927830097168/f
1. Create logins following the examples in the [figma design](https://www.figma.com/file/aM8i2nPES4Edf0EGXKhnwI/Autofill-(iOS)?node-id=2257%3A26485&t=7On7AyV4C09UdGwz-1)
2. Confirm the logins display and group correctly, including the More Options button 

**Copy updates (Translations TBD)**
1. Confirm copy update as per https://app.asana.com/0/0/1204129221412822/f
2. Confirm copy update as per https://app.asana.com/0/0/1203965896800930/f

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

amddg44 committed:
Autofill URLs now normalized to store only host:port

amddg44 committed:
Update to consider subdomains and ports for autofill suggestions

amddg44 committed:
Setting BSK to branch autofill/anya/subdomains-url-changes

amddg44 committed:
Temporary change for failing unit test due to subdomain work in progress

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into autofill/anya/subdomains-url-changes

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj

amddg44 committed:
Subdomains taken into account for grouping and sorting

amddg44 committed:
Unit test reset to correct value now that subdomain grouping work done

amddg44 committed:
Title behaviour changes

amddg44 committed:
Change to allow a preferred placeholder favicon letter to be displayed instead of default from domain

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into autofill/anya/subdomains-sorting

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj

amddg44 committed:
Fix text sizing issue

amddg44 committed:
Autofill login prompt changes for subdomain grouping + copy updates

amddg44 committed:
Linting fixes

THISISDINOSAUR commented on DuckDuckGo.xcodeproj/project.pbxproj:
```diff
@@ -4246,7 +4258,6 @@
 				859872221F5743AF00041CB8 /* FireAnimation */,
 				980891A122369ADB00313A70 /* FeedbackUserText.swift */,
 				985AAE4424899369007A43EC /* HomeScreenTransition.swift */,
-				1EF24234273BB9D200DE3D02 /* IntervalSlider.swift */,
```
Is this a deliberate removal? I don't see it in the file diff

amddg44 replied:
Good catch! I think something went awry with a merge commit from develop along with flattening all the PRs. Fixed

THISISDINOSAUR commented on DuckDuckGo.xcodeproj/project.pbxproj:
```diff
@@ -4263,7 +4274,6 @@
 				980891A82238504B00313A70 /* UILabelExtension.swift */,
 				9820FF4F2244FECC008D4782 /* UIScrollViewExtension.swift */,
 				982C87C32255559A00919035 /* UITableViewCellExtension.swift */,
-				319A37162829C8AD0079FBCE /* UITableViewExtension.swift */,
```
Same Q as above

amddg44 replied:
Fixed, as above

THISISDINOSAUR commented on DuckDuckGo/TabViewController.swift:
```diff
@@ -88,6 +88,10 @@ class TabViewController: UIViewController {
     internal lazy var featureFlagger = AppDependencyProvider.shared.featureFlagger
     private lazy var featureFlaggerInternalUserDecider = AppDependencyProvider.shared.featureFlaggerInternalUserDecider
 
+    private lazy var autofillUrlMatcher: AutofillDomainNameUrlMatcher = AutofillDomainNameUrlMatcher()
```
Is this reference to autofillUrlMatcher needed, or could it be initialised in the call to the AutofillWebsiteAccountMatcher init?

amddg44 replied:
Agree, not needed - have made that change

THISISDINOSAUR commented on DuckDuckGo/UserText.swift:
```diff
@@ -528,7 +528,13 @@ public struct UserText {
+        let message = NSLocalizedString("autofill.logins.prompt.partial.match.title", value: "From %@", comment: "Title for section of autofill logins that are an approximate match to the current website")
+        return message.format(arguments: type)
+    }
+
```
This reminds me we ought to get translations pretty soon

amddg44 replied:
Already in progress 🙂

THISISDINOSAUR commented on DuckDuckGo/AutofillLoginPromptHelper.swift:
```diff
@@ -0,0 +1,40 @@
+
+    static let moreOptionsLimit = 3
+
+    static func showMoreOptions(_ accountMatches: AccountMatches, limit: Int = Self.moreOptionsLimit) -> Bool {
```
I think this would be better as a question to make it clear it returns a bool (i.g. "shouldShowMoreOptions").

amddg44 replied:
Done!

THISISDINOSAUR reviewed (commented):
On the code side, same as BSK pr, I only have minor nits, and I love the testing!
Although this did remind me we need to get translations asap.

On the testing side most things worked as expected, but there were a few things that either seemed incorrect or I want to check: 

First test
"First test
Add a new login manually with a port, setting the Website URL to something like https://www.example.com:123
Save and check the website URL saved is example.com:123"
fill.dev saved as expected but for the other, www.example.com is displayed

Change to Favicon behavior 
"Confirm the placeholder letter is the first letter of the domain, not of the subdomain"
Seems to be incorrect, see screenshot
![Simulator Screen Shot - iPhone 14 Pro - 2023-03-10 at 18 50 12](https://user-images.githubusercontent.com/1093508/224403108-aa75f3e1-0f36-42f0-892e-19b2f337a5b1.png)


Re Changes to login prompt screen to illustrate partial subdomain matches
Doesn't offter to fill ones with a blank username, including ones that do have a password. Is that intentional?

amddg44 committed:
Small tidy up

amddg44 committed:
Renaming of function

amddg44 committed:
Update to AutofillLoginDetailsHeaderView to use preferredFakeFaviconLetter

amddg44 committed:
Linting fix

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into autofill/anya/subdomains-sorting

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj

amddg44 committed:
Fix due to changes to StorageCache

amddg44 committed:
Removed unused tld variable

amddg44 committed:
Updated to use tld from StorageCache

amddg44 committed:
Updated for changes to BSK function names

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into autofill/anya/subdomains-sorting

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj

amddg44 committed:
Updated Autofill translations

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into autofill/anya/subdomains-sorting

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into autofill/anya/subdomains-sorting

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj

amddg44 committed:
Autofill prompt sheet behavior change (#1568)

Task/Issue URL: https://app.asana.com/0/0/1204122171012219/f
Tech Design URL:
CC:

Description:
Changes the behavior of the Autofill prompt sheet to be the following:

Initial sheet supports medium detent only (cannot be dragged up to full screen)
Tapping "more options" should close the current sheet and open a new sheet with only a large detent (cannot be dragged down to medium)

amddg44 committed:
Fix for Autofill toast text incorrect when deleting login (#1569)

Task/Issue URL: https://app.asana.com/0/0/1203782162439772/f
Tech Design URL:
CC:

Description:
This is a fix for when a login without a title is deleted, the correct toast message is displayed.

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into autofill/anya/subdomains-sorting

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj

amddg44 committed:
Bump to BSK release 53.1.0

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into autofill/anya/subdomains-sorting

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
